### ISSUE TITLE
Custom materials are not saved 

### ISSUE THREAD

**Issue by marcopiraccini** [Ultimaker/Cura#1104]
Cura 2.3.1 (PPA) on Ubuntu 16.04
I duplicate a generic material, then I edit it, changing name and stuff.
When I close and reopen Cura, the material I created is disappeared. 

Also, if I export materials, the import seems not to work

**Comment by nallath** [comment]
Could you add your logs?

**Comment by marcopiraccini** [comment]
Sure. Actions done: 
- Started Cura
- Materials -> Duplicate PLA
- Edited the duplicated PLA, called "PLA - NEW" and set the diameter to 1.75mm
- Closed Cura => `
[log_cura_1.txt](https://github.com/Ultimaker/Cura/files/570827/log_cura_1.txt)

- Reopened Cura. Reopened Materials, "PLA - NEW" NOT was present =>
[log_cura_2.txt](https://github.com/Ultimaker/Cura/files/570828/log_cura_2.txt)

**Comment by Ghostkeeper** [comment]
The logs make no mention of any "PLA - NEW" :(

**Comment by BerndJM** [comment]
@Ghostkeeper: The log didn't refer to the given name but to a kind of filename and counter `....#2`.
I can confirm what @marcopiraccini wrote, trying to help to clearify things I made a "commented logfile" where you can see the steps I've done at the different moments in the log, maybe it helps.
[cura.log.commented.txt](https://github.com/Ultimaker/Cura/files/572352/cura.log.commented.txt)

**Comment by BerndJM** [comment]
o.k. got it (I think) - compared the generated file with the material-files shipped with cura
@marcopiraccini could you please crosscheck this:
in your new generated material file in ~.local/share/cura/materials should be a line:
`<compatible>True</compatible>`
just delete this line and save the file and it should be visible within curas material-select

**Comment by marcopiraccini** [comment]
Yes! removing `compatible` the material is now visible. 
I guess that this is a bug on saving...

**Comment by BerndJM** [comment]
thanks for confirmation.
Yeah a bug in saving ... or in reading ... that's the question ;-))
@Ghostkeeper it's your turn now ...

**Comment by Ghostkeeper** [comment]
That's some nice debugging, guys! We'll take a shot at this then.

**Comment by MabuClou** [comment]
Ok. It is anoying. This line is always readded. So i always have to delete this line `<compatible>True</compatible>` 
before starting Cura.

Also, this line seems to be added when duplicating a material. So the new material will not even show up in the manager, while there is actually a file created in the directory.

And also, it is not possible to edit the fields about filamentcost and spoolweight. Is there a documentation about the XML-Structure of this file?

And when starting the Material Manager, then it does not show all my custom materials. I have used brand "ColorFabb", so the should stay on the top of the list. But there is only one line. I have to scroll down the list, before all the materials are available on the top of the listing.

**Comment by Ghostkeeper** [comment]
The filament cost and weight editing was a feature that didn't get done in time for 2.3. It'll be available later. For now the field only displays the cost and weight information (if available).

**Comment by dWoolridge** [comment]
Please note this problem is also active on Windows 7 64-bit.

**Comment by dWoolridge** [comment]
Any idea when this will be fixed?  I really would like to switch to 2.3.1 (from 15.02.1), but this is preventing me from making the upgrade.  I'm really looking forward to seeing how coasting and retraction priming can improve print quality.

**Comment by carlosjln** [comment]
I didn't know about this issue until now that I decided to use Process Monitor and do my own debugging, I guess I'm on the waiting list as well :)

**Comment by Ghostkeeper** [comment]
We've had some changes last week that were aimed at other things, but also seems to have fixed this issue. I can't reproduce it any more with these reproduce steps (which used to reproduce it):

* Add an Ultimaker 2+ machine.
* Go into Manage Materials.
* Duplicate the TPU material.
* Rename the TPU material to TUTURU-TPU.
* Close Cura and restart it.
* Check if the TUTURU-TPU material is present.

So I hope it's fixed, but I'd like to keep this one open at least until the beta of 2.4 to see if it still occurs with other materials or something.

**Comment by StefanKi** [comment]
Unfortunately the problem isn't fixed for me.
I am using ppa:thopiekar/cura-master.
Should it already be fixed there?

**Comment by BerndJM** [comment]
No, it's not (fully?) fixed at the moment.
You can delete the line mentioned above from the file and the definition is visible in Cura.
But a soon as you edit the material definition (in Cura) and it's saved, this line will be written again in the file :-(

**Comment by Ghostkeeper** [comment]
I still can't reproduce it with the steps I described on December 5th.
Is there something different you're trying? Could you try again with the 2.4 beta from the Ultimaker website?

**Comment by StefanKi** [comment]
Since i am using ppa:thopiekar/cura-master i don't think, that using the version from the website would be more up to date. Lately all configuration was lost and i had to redo all confugurations - but the problem persisted.

I will retry it with the latest ppa:thopiekar/cura-master version and post the results the day after tomorrow.

**Comment by Ghostkeeper** [comment]
The (beta) version on the website was made last Thursday.

**Comment by dWoolridge** [comment]
I'm not sure if this contributes in some way to the problem but here's what I think just happened...
1)  Downloaded and installed Cura 2.4 Beta
2)  Created new Material--I'm not 100% sure, but I think it still existed after closing and re-opening Cura 2.4 beta
3)  Opened Cura 2.4 Beta
4)  Opened Cura 15.02.1 at the same time-to see what my old settings were as I entered them into Cura 2.4
5)  Problems with "materials not saved" re-appeard

**Comment by dWoolridge** [comment]
Please disregard my previous post.  Opening Cura 15.02.1 along with 2.4.0 Beta doesn't appear to have anything to do with the problem.  The following steps show that the "<compatible>True</compatible>" line is still added to the file any time a custom material is edited.

1. Uninstalled Cura 2.4.0 Beta
2. Deleted all files from C:\Users\....\AppData\Local\cura\materials
3. Re-installed Cura 2.4.0 Beta (on my E: drive, not the C: drive - otherwise default installation)
4. Opened Cura 2.4.0 Beta
5. Duplicated the "Generic ABS" material
6. Edited the new changing:  Display Name, Brand, Color, Color Description, Diameter to 1.7, Print temp to 220C, Retraction to 3mm, Retraction speed to 25
7. Clicked "Close" on the material editor
8. MATERIAL IS THERE (in Cura's material list)
9. Closed Cura 2.4.0 Beta
10. Open Cura 2.4.0 Beta
11. MATERIAL IS GONE (Doesn't show in Cura's material list)
12. Closed Cura 2.4.0 Beta
13. Removed  "<compatible>True</compatible>" line from material file
14. Open Cura 2.4.0 Beta
15. MATERIAL IS THERE (in Cura's material list)
16. Selected the material
17. Closed Cura 2.4.0 Beta
18. Open Cura 2.4.0 Beta
19. MATERIAL IS THERE (in Cura's material list)
20. Edited material (added a character to the name)
21. Clicked "Close" on the material editor
22. Closed Cura 2.4.0 Beta
23. Checked material file and the "<compatible>True</compatible>" line is back
24. Open Cura 2.4.0 Beta
25. MATERIAL IS GONE (Doesn't show in Cura's material list)26.

**Comment by StefanKi** [comment]
Just tried ppa:thopiekar/cura-master 1:2.4.99-master~201612281202~rev1647~pkg130~ubuntu16.04.1
Not working - materials are not saved.

What can I do to help you finding out the problem?
Should I send you my configs? (If yes - which ones?) Maybe the log from the terminal?

**Comment by StefanKi** [comment]
I think i have found out something very useful!?

I added an Ultimaker 3 Extended just for fun, and suddenly had options for 2 extruders?
And now for each of this extruders suddenly i could see my custom materials!?!?

Now i can switch between 2 Printers:
1. my custom printer
2. Ultimaker 3 Extended

When switching to my custom printer, all custom materials are gone.
When switching to the Ultimaker 3 Extended, all custom materials are there.

How could that be? I didn't find any option for more than 1 extruder for my custom printer?
Maybe you only have to activate the more than one extruder options for the custom printers and the custom material bug would be gone...?

**Comment by Ghostkeeper** [comment]
Did you make that custom printer with that "Custom FDM Printer" menu, or did you create a .def.json file manually?

**Comment by StefanKi** [comment]
I did it over the menue.

Is there any docu about the json files with the 2 (or more?) extruders?

**Comment by Ghostkeeper** [comment]
Not really. To produce a multi-extrusion printer definition, it's advisable to copy and rename all files related to the Ultimaker 3. Or Punchtec if you don't want to go through the trouble of creating all the quality profiles. But even I still make mistakes when I have to create a new multi-extrusion printer definition. It's quite complicated.

**Comment by dtrazzi** [comment]
getting same problem here. Have a printer added as a Custom FDM, and added materials disappear once the program is closed and re-opened. Using Cura 2.3.1

**Comment by agarthand** [comment]
Hope I can contribute to this issue...

Facing the same problem of disappearing Material (created via duplicate existing) and restarting Cura.
I'm using a Custom FDM printer and tried this with the Cura 2.4beta.

The interesting thing is that after creating additional printer instance of type "Ultimaker 3" and activate this I can see all the hidden material in (both) extruder material view for this printer. But as soon as I change back to my existing printer it's gone again.

Would love to see this fixed but for the time being I might have to manually customize the compatible line every time. Probably need to write a short script to automatize it ;)

**Comment by matthock** [comment]
I'll add another voice on this. I am using 2.4beta, and I am seeing the exact same issue as described above. I tried the same test sequence as @dWoolridge and produced the same results.

**Comment by Ghostkeeper** [comment]
We had marked this issue as fixed in 2.4, but I just put it back so that someone will take another look at it for version 2.5.

**Comment by Ghostkeeper** [comment]
Our QA testers say they can't reproduce this on Windows or OSX. They tried with UM3 and UM3 Extended, with various materials.

**Comment by StefanKi** [comment]
Yes it is working with the UM3.
But it is NOT working with the Custom Printers.
I don't know if the OS is important - I don't have any Windows or OSX to test that.

**Comment by marekzehra** [comment]
Just tested OSX version 2.4.0-BETA2. Not working for me.
Steps to reproduce:

1. Add custom printer (In my case Kossel Mini)
2. Select Material->Manage materials...
3. Duplicate Generic PLA (rename, add brand, whatever...)
4. Repeat step 3 to create my set of custom materials (PLA, different colors, sizes, brands)
5. Quit Cura
6. Start Cura again
7. Select Material

Materials gone!

**Comment by Ghostkeeper** [comment]
Thanks. The QA testers report that it does reproduce for other printers than UM3 and UM3E.

**Comment by ntoff** [comment]
Same issue here on 2.4 beta (mendel90 printer) but if I manually create a material everything is fine, they only don't work if I use cura to generate the material. I've taken to just copy / pasting the material files and manually editing the contents, this results in usable material profiles.

Here's one of my custom materials that *does* work https://gist.github.com/ntoff/495604311240980108815c6bff6a2e54 I started generating random uuid's for them but found it doesn't really matter for my printer / usage (I guess only ultimakers make use of that?)

**Comment by Chase-san** [comment]
Same thing here, removing the compatible tag from the file makes it work. For now I am setting the files to read-only, otherwise they revert (and don't show up in the menus).

**Comment by Ghostkeeper** [comment]
Our bug tracker says that this is on schedule for fixing in 2.4, 2.5 ánd 2.6! I guess they find it really important... Or Jira is just a bit of a mess. 2.4 is locked, so I'd expect 2.5 then.

**Comment by StefanKi** [comment]
For me the problem seems to be fixed in the latest ppa:thopiekar/cura-master.
Maybe otherones can confirm it?

**Comment by MarcusWolschon** [comment]
I too can see my "nGen" material after upgrading from Cura 2.3 to 2.4

**Comment by Ghostkeeper** [comment]
That's strange, because the issue is still unfixed according to our QA testers.

**Comment by bubbapizza** [comment]
This problem still exists.  I haven't used Cura since the numbering scheme was changed and I will guarantee materials do not save properly if you install it on a brand new ubuntu 16.10 box and configure a custom machine.  I just found this thread by accident because I KNOW I configured some materials and they didn't get saved when I closed the app and restarted it.

Sure enough the files are in .local/share/cura/material and they have this line in them:

<setting key="hardware compatible">yes</setting>

**Comment by bubbapizza** [comment]
Ohh, wait a sec, there's something wrong here and I think it's because there's no way to create new materials other than choosing "duplicate".  Who designed this anyways?

**Comment by bubbapizza** [comment]
Ok, materials management IS SERIOUSLY BROKEN on ubuntu 16.10 and cura 2.4.  Brand new install, never used cura 2.4 before, so matter what changes you make, no materials ever get saved.  It's a mess.

**Comment by ntoff** [comment]
@bubbapizza Try manually managing your materials, I agree the inbuilt system is far too broken to bother with, use this as a template: https://gist.github.com/ntoff/495604311240980108815c6bff6a2e54 and manually create your materials with a good text editor (or nano / vi if you prefer command line stuff). I've been making my own manually on windows and it works great.

**Comment by fieldOfView** [comment]
> Who designed this anyways?

'sup?

**Comment by bubbapizza** [comment]
Why is there no 'add' button for materials?  This is just a common-as-dirt CRUD database.  You need an add/update/delete/next/previous view to show the details of each material, then a browse view where you list the materials like a spreadsheet.  The "duplicate" button should be an option shown AFTER you choose "add".

**Comment by StefanKi** [comment]
Feel free to change everything you want. It's open source...

**Comment by bubbapizza** [comment]
This is true

**Comment by Ghostkeeper** [comment]
I agree that it would be nice to have an "Add" button. We only implemented "Duplicate" because it was easier because you have a GUID and some profiles to start with, and due to time constraints.

However that is a different issue and should not be discussed in here.

**Comment by LFATCGS** [comment]
Uggghh jeez, I was wondering why my printer wasn't working. Cura displays myXyzPLA as the material but when I click manage material - there is no myXyzPLA anymore! So it has been printing everything with 2.85mm instead of 1.75mm diameter and who knows how many other settings were lost - probably all of them. Poor form.

**Comment by phaerus** [comment]
Can confirm issue still present in Windows 10 Professional, Cura 2.4. Using a custom printer, not Ultimaker. Removing the `<compatible>True</compatible>` line from the material files in `%localappdata%/cura/materials/` lets me view the material in the UI.

**Comment by nallath** [comment]
A machine can be compatbile "in general", but a machine can have inter machine differences (in most cases; nozzle size or shape). We call those "variants". 

It can be the case that the machine in general supports it, but a single one of those variants doesn't. In that case you define the compatibility of that variant to false.

**Comment by Kadah** [comment]
Ran into this as well. I'm new to Cura and 3d printing in general. I've barely touched Cura 2.x so far as I found the UI very difficult to navigate and use over all the other frontends I've tested.

**Comment by BagelOrb** [comment]
Is this issue still not fixed?

**Comment by fieldOfView** [comment]
I think it is fixed in at least 2.5b2. At least I could no longer reproduce it when I tested it a week or so ago.

**Comment by seantapscott** [comment]
Custom materials work for me.  I can duplicate and edit them. 

I also checked that custom material export/import works (although the open/save dialog has a weirdly set default folder lookup).  Also, strangely, the display name of the imported material is the name of the file, and Cura hangs whenever I edit custom materials (probably a different issue entirely). 

- Using 2.4.0

**Comment by Ghostkeeper** [comment]
Still seem to get this bug on Master though when I use the steps described in #1687.

**Comment by MarcusWolschon** [comment]
Using 2.4.0 I created a new material "Colorfabb CF20".
Now I have that material still currently selected for the printer "Ultimaker 2ex 0.4mm nozzle" but it and it's category are no longer in "manage materials" and I can not e selected when changing to a different printer such as "Ultimaker 2ex 1.0mm nozzle".
The custom Material "Colorfabb nGen" that I used daily for many month before has completely disapeared from Cura.

**Comment by MarcusWolschon** [comment]
After the update to Cura 2.5 all custom materials are still there... but each has 20-30 copies that didn't exist before the update. No keyboard navigation in the list and no way to select multiple  with shift. So I have to delete each one of them and it takes serveral seconds to delete a single entry each time.

**Comment by nallath** [comment]
The cost + weight fillament is not a bug; You make duplicates. That means that they share GUID (so they are about the same material). A material has a single cost.

**Comment by fieldOfView** [comment]
A workaround for the cost+weight thing is to (manually) edit the GUID in the XML file to something unique. Cost and weight are not stored in the XML file, but in the Cura configuration, and they are stored by the value of the GUID.

I will see if I can get around to adding a way to "unlink" the GUID in the cura interface.

**Comment by nallath** [comment]
Having a new GUID is making a new material. If you want to go there, I think its best to just add a "new" button + GUI to create a new material.

**Comment by fieldOfView** [comment]
The thing is that people have been using the "Duplicate" button for a couple of months now, so we also need to solve the problem caused by that.

I'm thinking of detecting a duplicate GUID and adding a message on top of the details pane that says "this material is currently linked to "Generic PLA". Would you like to unlink it?", where unlink would be a "hyperlink"/button that creates a new GUID.

**Comment by nallath** [comment]
You're right, but i'd also like to prevent people from getting more used to the incorrect workflow of duplicating a material to get a new one.

**Comment by MarcusWolschon** [comment]
"A workaround for the cost+weight thing is to (manually) edit the GUID in the XML file to something unique."
As I already wrote. This does not work.

Exporting a material, deleting the original, then changing the GUID in the exported version, then importing it again, then adding cost+weight again manually because of #1715 seems to be a workaround.  But not a workflow you can offer to any user.

**Comment by fieldOfView** [comment]
Don't import/export

* open the configuration folder (via the help menu), 
* close cura
* find your material in the materials folder
* edit the guid 
* restart cura

The material is now no longer linked to the one you have copied it from.

No, this is not a great solution, this is a _workaround_. For now.

**Comment by MarcusWolschon** [comment]
I did that, gave each custom material a different last digit in the GUID and it wrecked havoc.
Maybe these GUIDs where referenced somewhere? As a current material in the 3 custom printers perhaps?
(I didn't check what GUID format version was used by Cura in the third block. Maybe I broke some checksum.)

**Comment by nallath** [comment]
GUID is just seen as a string (that has to be unique). "zomg" should work as GUID

**Comment by fieldOfView** [comment]
See #1757

**Comment by ftoons** [comment]
I'm facing the issue of disappearing  profiles in Cura 2.6.2. in Windows 10.
My first install was 2.5. I made some costum material that worked well. Upgraded to 2.6 beta and profiles were there. Upgraded to 2.6.1 when it came out and no problems again.
Then 2.6.2 came out and after upgrading the profiles come and go occasionally! no specific pattern or trigger that makes them appear or disappear, except that some randomly appear when I change the material.
I hope I can get some help with that.
Thanks.

**Comment by Ghostkeeper** [comment]
We made no change in that aspect for 2.6.2. Actually 2.6.2 was basically just an update for MacOS to have the correct CPU architecture target, and a useless update of the firmware builds.

That said, I can't reproduce the issue you're reporting on my computer.

**Comment by ftoons** [comment]
Thanks for your reply.
So it seems the problem in my case is that the profiles disappear based on selected material, even if on extruder 2.  
Here is one way I can reproduce the problem:
If extruder 1 is AA and has a generic PLA spool and I choose (Ultimaker black PLA (or any ultimaker filament)) in Cura
AND ASLO
Extruder 2 is AA core and has Ultimaker PLA spool in it, and I select material as PVA in Cura (Not compatible message shown), I can only see very few of my profiles. The original profiles disappear and I  only see two of the ones i made.

Now I can get all the profiles back by changing materials and cores in Cura, but I don't know based on what. I don't recall I was told by cura tutorial or a popup message or something that profiles are core / material dependent and i don't know precisely makes them show up or hide from the menu.

Thank you for your support

**Comment by Ghostkeeper** [comment]
Ah yes. We only show quality profiles with a quality type that all extruders can support. There is a bit of a weird emergent result there because the "Not Supported" profiles are also profiles like any other and they have their own quality type. As a result, the AA/PVA combination thinks only profiles with 0.1mm layer height are supported. I expect that this will change when we implement profiles depending on combinations of materials, which is scheduled for the end of the year. We'll have to think of better fall-backs then.

PVA is actually forbidden on an AA nozzle since it damages the nozzle. Other materials are only supported on AA but still printable on BB.

This has nothing to do with custom materials not being saved though so if you want to continue this conversation, please make a separate issue for it.

**Comment by Ghostkeeper** [comment]
This was fixed in version 3.0, if I recall correctly.

**Comment by aenertia** [comment]
Creating and Duplicating Materials in 3.3.1 (Linux appimage) fails. The XML's get generated however, it appears that the machine type that gets added (on either a duplicate or create material) event only include ultimaker product strings, which then hides the material (My Machine is a custom FDM) - changing the filament size to say 1.75mm is enough to trigger this. Attached an example xml of what happens. These materials disappear the moment they are created with the UI and makes material management pretty much useless. This has been happening for at least the last 2 major releases for me. (Always on Linux with Appimage)
[custom_material+%232.xml.fdm_material.txt](https://github.com/Ultimaker/Cura/files/2087546/custom_material.232.xml.fdm_material.txt)

**Comment by Ghostkeeper** [comment]
The material diameter may make a material disappear if it's incompatible with your current printer. Look into the extruder settings in the machine settings dialogue to see what material diameter it needs.

If a material disappears for this reason, Cura should be showing a warning message.

I'm unable to reproduce your bug with the following steps in 3.4.1:
* Add Custom FDM Printer.
* Go to Material Manager and duplicate PLA.
* Rename the duplicated material to BLA.
* Restart Cura.

Before and after restarting Cura, the selected material is still BLA, and BLA still appears in the list of materials.

**Comment by magicalbeard** [comment]
This bug is currently present in 4.5.0 - all my duplicated materials are being updated at once when I change the material cost.

**Comment by Ghostkeeper** [comment]
If you don't want the material cost to be shared between multiple profiles, click the "unlink" button right above it. These are linked to allow you to have different sets of settings for the same spool of material.

**Comment by MarcusWolschon** [comment]
Why would anyone do that? It's material properties. If you want to deviate from them, you can still override them in your print profile.

**Comment by magicalbeard** [comment]
![image](https://user-images.githubusercontent.com/3375292/80115253-872e6a80-8552-11ea-8cc7-908968d53d73.png)

Where is the "unlink button"?

**Comment by Ghostkeeper** [comment]
> Why would anyone do that? It's material properties. If you want to deviate from them, you can still override them in your print profile.

Except for the diameter, these are not settings and so they can't be stored in your custom print profile. They are metadata entries for material profiles. They don't have any effect on the slice either.

> Where is the "unlink button"?

If the material is not linked to any other material, there is no unlink button. If it is linked, it will be between the properties and the description:

![Screenshot from 2020-04-28 16-38-50](https://user-images.githubusercontent.com/2448634/80500782-e46d5600-896e-11ea-8a20-fd57bcbb8235.png)

If you duplicate a material profile, your material will be linked to the original. If you create a new material profile, it won't be linked to anything (but under water it will start with the properties of PLA).

**Comment by magicalbeard** [comment]
@Ghostkeeper This is where I'm really confused - I have duplicated the Generic PLA material, cannot unlink it (because there is no button) and it still shares those 4 metadata values between all of them. I can, however, edit anything under "Print Settings" and those changes are unique to that duplicate. 

I have no "unlink material" button on any of my duplicated materials. This means that if I want to have two identical materials, differing only in their material cost/density/diameter or weight, I would have to create them from scratch and cannot inherit any attributes from an existing material. There are many manufacturers that sell different colors of identical base filament at different prices.

Im running Cura 4.5.0 on windows 10, I've tried many different ways of making new materials and cannot get that "unlink" button to show up.

**Comment by Ghostkeeper** [comment]
> I can, however, edit anything under "Print Settings" and those changes are unique to that duplicate.

Yes those are not linked, and really the point of being able to duplicate materials. You can edit the print settings and have different profiles for the same spool. This is not a bug.

However that you see no "unlink material" after duplicating, and that the cost and weight are still linked seems like a bug to me. Only I'm unable to reproduce that. Maybe there is a bug that the button is not showing?

The list of linked materials is constructed here:

https://github.com/Ultimaker/Cura/blob/8386d094d3bc9226754b8e2d4b6728735283380b/resources/qml/Preferences/Materials/MaterialsView.qml#L34-L50

The button is shown if that list is not empty.

This code is quite noodly though. I wouldn't be surprised if there were some update bugs in there. It's been on our list to refactor for a while.
